### ISSUE TITLE
issue with the kernel expected when running Elyra pipeline

### ISSUE THREAD

**Issue by pacospace** [elyra-ai/kfp-notebook#88]
**Describe the issue**
I created a new kernel (e.g. `python3new` ) and added it to the notebook kernel, therefore it is part of the notebook metadata and it can be found from command:
```
[1000600000@jupyterhub-nb-fmurdaca-40redhat-2ecom python3]$ jupyter kernelspec list
Available kernels:
  python3-test    /opt/app-root/src/.local/share/jupyter/kernels/python3-test
  python3new      /opt/app-root/src/.local/share/jupyter/kernels/python3new
  python3         /opt/app-root/share/jupyter/kernels/python3
```

Then when submitting an Elyra pipeline the kernel selected is not found.

Looking at the code my understanding is that on the image used as base runtime for the step in the pipeline that kernel name actually does not exists, so if the kernel name from notebook metadata is selected it will fail as it cannot find kernel.json file (https://github.com/elyra-ai/kfp-notebook/blob/a1f1a9d4ba21b1d2b664afce440095f1f0912cd1/etc/docker-scripts/bootstrapper.py#L413). Switching back to `python3` kernel name in notebook metadata let the step run without problems on the runtime image.

Is it an expected behaviour? I would expect that the new-kernel name (the dependencies are provided by the runtime image used for the step so there is no problem) which is part of the notebook metadata does not cause failure in the step. Moreover, the issue identified is a bit ambigous because the kernel.json for the actual kernel name is the file missing and not sure why also the notebook is missing.

**To Reproduce**
1. Create a new kernel `python3new` with `jupyterlab-requirements` [[2]] for example
2. check the kernel is available `jupyter kernelspec list`
3. Assign the kernel to the notebook
4. Run an Elyra pipeline with runtime image that is compatible.

**Screenshots or log output**
```
[I 17:02:58.610] 'elyra-aidevsecops-tutorial-0304170220':'download_dataset' - dependencies processed (1.161 secs)
[I 17:02:58.611] 'elyra-aidevsecops-tutorial-0304170220':'download_dataset' - executing notebook using 'papermill download_dataset.ipynb download_dataset-output.ipynb' 
[D 17:02:59.327] Native kernel (python3) available from /opt/app-root/src/jupyter-work-dir/python3/ipykernel/resources
[E 17:02:59.327] Unexpected error: <class 'FileNotFoundError'>
[I 17:02:59.457] 'elyra-aidevsecops-tutorial-0304170220':'download_dataset' - converting from download_dataset-output.ipynb to download_dataset.html 
Traceback (most recent call last):
  File "bootstrapper.py", line 341, in execute
    kernel_name = NotebookFileOp.find_best_kernel(notebook)
  File "bootstrapper.py", line 415, in find_best_kernel
    with open(os.path.join(file, 'kernel.json')) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/opt/app-root/src/jupyter-work-dir/python3/ipykernel/resources/kernel.json'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/app-root/src/jupyter-work-dir/python3/nbformat/__init__.py", line 138, in read
    buf = fp.read()
AttributeError: 'str' object has no attribute 'read'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "bootstrapper.py", line 594, in <module>
    main()
  File "bootstrapper.py", line 584, in main
    file_op.execute()
  File "bootstrapper.py", line 356, in execute
    NotebookFileOp.convert_notebook_to_html(notebook_output, notebook_html)
  File "bootstrapper.py", line 374, in convert_notebook_to_html
    nb = nbformat.read(notebook_file, as_version=4)
  File "/opt/app-root/src/jupyter-work-dir/python3/nbformat/__init__.py", line 140, in read
    with io.open(fp, encoding='utf-8') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'download_dataset-output.ipynb'
```

**Expected behavior**
No failures in the Elyra pipeline.

**Deployment information**
Describe what you've deployed and how: [[1]]
 - Elyra version: 2.* [[1]]
 - Operating system: linux [[1]]
 - Installation source: [[1]]
 - Deployment type: Open Data Hub on Openshift, Elyra image spawn from JH and Kubeflow + Tekton as engine for pipelines run. [[1]]

[1]: https://github.com/thoth-station/elyra-aidevsecops-tutorial/blob/master/README.md#environment-required
[2]: https://github.com/thoth-station/jupyterlab-requirements

**Comment by kevin-bates** [comment]
Hi @pacospace - @akchinSTC and I are looking into this and we've been able to reproduce it using your image when `CRIO_RUNTIME=true`.  Interestingly enough, it does not reproduce when the default container runtime is used.

I set my notebook file to reference a non-existent kernel named `python-foo` and I get this in the KFP logs for that node:
```
[D 22:59:04.233] Found kernel python3 in /opt/app-root/share/jupyter/kernels
[I 22:59:04.234] Matched kernel by language (python), using kernel 'python3' instead of the missing kernel 'python-foo'.
```

However, when I run the same pipeline against elyra with `CRIO_RUNTIME=true` I get the following...
```
[D 23:16:44.514] Native kernel (python3) available from /opt/app-root/src/jupyter-work-dir/python3/ipykernel/resources
[E 23:16:44.514] Unexpected error: <class 'FileNotFoundError'>
[I 23:16:44.655] 'py-foo-crio-0304151557':'node1' - converting from node1-output.ipynb to node1.html 
Traceback (most recent call last):
  File "bootstrapper.py", line 341, in execute
    kernel_name = NotebookFileOp.find_best_kernel(notebook)
  File "bootstrapper.py", line 415, in find_best_kernel
    with open(os.path.join(file, 'kernel.json')) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/opt/app-root/src/jupyter-work-dir/python3/ipykernel/resources/kernel.json'
```
I will address the (failing) support for the "native" kernel, but we also think there's something amiss between a default KFP with CRIO_RUNTIME vs. ODH-based configurations from RedHat - where the CRIO stuff was first qualified.

Are you able to run elyra with `CRIO_RUNTIME=false` (or unset)?

**Comment by akchinSTC** [comment]
While looking at this issue, I may have come across something else not quite right
When `CRIO_RUNTIME` is set, the notebook operator will explicitly sets the `PYTHONPATH` to `/opt/app-root/src/jupyter-work-dir/python3/` via container methods and does not prepend the existing `PYTHONPATH`.

**Comment by kevin-bates** [comment]
Thanks @akchinSTC.  I've just created PR #89 to address the faulty handling of the "native kernel" - which should allow this issue to proceed.  

I recommend we open a new issue to address the `PYTHONPATH` portion of this.

**Comment by pacospace** [comment]
> **TL;DR:** We're not handling "native kernel" fallback logic properly and your image introduced a scenario that both my test with your image or any other scenarios had yet to uncover. That said, if your pipeline is dependent on your custom kernel configuration, you will need to ensure that that kernelspec is resident in your image and discoverable via the best-match logic.

At the moment what we do is:

- create kernel with dependencies using jupyterlab-requirements, (so dependencies are stored in the notebook metadata, togheter with runtime environment used)
- test notebook locally
- store notebook and dependencies on the github repo
- create a release and an image (compatible with the runtime environment used) that can be used specifically to run that notebook as base.

Therefore in principle, the use of the default kernel created in the image once Elyra installs all requirements is "working" for our case, however it is stil weird if a custom kernel could not be used. 

What if Elyra could:

- use dependencies stored in the notebook metadata
- take them, togheter with kernel name, to create a virtualenv in the base image used for the step and install the dependencies there, so that they do not create conflict with the requirements coming from kfp-notebook library?

these steps could be done using thamos library (https://github.com/thoth-station/thamos) for example, which also includes micropipenv (https://pypi.org/project/micropipenv/)
- use runtime environment information stored in the notebook metadata to verify the base image is compatible and you are not using fedora 33 when your notebook has been created using fedora 32, and warn the user that should use a proper base image.

- run kfp-notebook logic with papermill using the notebook and the proper kernel?


WDYT? @kevin-bates @akchinSTC 

> 
> Hmm - this is a bit odd. Here's what I've done...
> First, I attempted to run `jupyter kernelspec list` directly against your image and got the expected result...
> 
> ```
> $ docker run -it --rm quay.io/thoth-station/elyra-aidevsecops-tutorial:v0.5.0 jupyter kernelspec list
> /usr/bin/container-entrypoint: line 2: exec: jupyter: not found
> ```
> 
> Next, I extended the image with the previously linked `elyra-requirements.txt` file using this Dockerfile:
> 
> ```
> $ cat Dockerfile 
> FROM quay.io/thoth-station/elyra-aidevsecops-tutorial:v0.5.0
> 
> COPY requirements-elyra.txt /tmp
> RUN python -m pip install -r /tmp/requirements-elyra.txt
> ```
> 
> and this `docker build` command...
> 
> ```
> $ docker build . -t elyra-aidevsecops-w-reqs
> ```
> 
> It turns out a plain installation of `ipykernel` does install a kernespec...
> 
> ```
> $ docker run -it --rm elyra-aidevsecops-w-reqs jupyter kernelspec list
> Available kernels:
>   python3    /opt/app-root/share/jupyter/kernels/python3
> ```
> 
> and the kernelspec directory contains a properly formatted `kernel.json` file...
> 
> ```
> ~/dev/docker/elyra-custom$ docker run -it --rm elyra-aidevsecops-w-reqs cat /opt/app-root/share/jupyter/kernels/python3/kernel.json
> {
>  "argv": [
>   "python",
>   "-m",
>   "ipykernel_launcher",
>   "-f",
>   "{connection_file}"
>  ],
>  "display_name": "Python 3",
>  "language": "python"
> }
> ```
> 
> I also get the same directory using code that is used in the best-kernel-match logic (just wanted to make sure the CLI command didn't produce anything unusual):
> 
> ```python
> from jupyter_client.kernelspec import KernelSpecManager
> 
> def best_kernel():
>     kernel_specs = KernelSpecManager().find_kernel_specs()
>     for name, file in kernel_specs.items():
>         print(f"{name}: {file}")
> 
> if __name__ == "__main__":
>     best_kernel()
> ```
> 
> However, another clue is this DEBUG statement:
> 
> ```
> [D 17:02:59.327] Native kernel (python3) available from /opt/app-root/src/jupyter-work-dir/python3/ipykernel/resources
> ```
> 
> This DEBUG is printed from the [`KernelSpecManager().find_kernel_specs()`](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/kernelspec.py#L171-L178) method when the [`NATIVE_KERNEL_NAME`](https://github.com/jupyter/jupyter_client/blob/master/jupyter_client/kernelspec.py#L24) is not already in the results list. Since that value is `python3`, this implies that, for whatever reason, the installation of the elyra-requirements did NOT create a valid kernelspec. And, since the 'file' element of the result set points at the [`RESOURCES`](https://github.com/ipython/ipykernel/blob/master/ipykernel/kernelspec.py#L21) directory of where `ipykernel` is installed, the matching logic doesn't handle that (and it should). In essence, whenever the "native kernel" is used as a fallback, the `kernel.json` is essentially created on the fly. As a result, when the bootstrapper encounters this condition (which I would express as a) `kernel.json` is missing and b) the basename of the directory is `resources`) then we should treat this as a python-language kernel for matching purposes. If that's a language match, we're done - else move on.
> 
> What I don't understand is why the creation of the proper kernelspec worked in my scenario, but not yours. I suspect it might have something to do with where python is located in the path etc. Since the Native kernel is found in bowels of `jupyter-work-dir` under `/opt/app-root/src/jupyter-work-dir`, I suspect the jupyter path logic used by Jupyter code is not finding the non-native (and expected) kernelspec definition and only the fallback - or the non-native spec if failing to be installed during the requirements processing.
> 
> In fact, I see that my experiment probably isn't valid since we adjust the [working python lib directories](https://github.com/elyra-ai/kfp-notebook/blob/master/kfp_notebook/pipeline/_notebook_op.py#L134-L140) via the Notebook operator.
> 
> At any rate, since the change for this issue needs to come from the kfp-notebook repo, I'm going to transfer this there.
> 
> cc: @akchinSTC

**Comment by kevin-bates** [comment]
> create kernel with dependencies 

What exactly does this mean?  Using your "custom kernel" from above - `python3new` - in what manner are you altering the directory `/opt/app-root/src/.local/share/jupyter/kernels/python3new`?  Could you provide its `kernel.json` and describe any other alterations, additional files, added into that directory's hierarchy?

Regarding propagation of the notebook's requirements metadata into the container, there's a bit of a chicken/egg issue here.  I wonder if we could introduce a "pre-step" during the node's processing in the bootstrapper sequence that extracts these custom requirements from the notebook's metadata and ensures their installation prior to the notebook's execution?

I'm not sure if this step would be done against its own virtual env or not.  I also don't see how we could entertain the kind of version checking/warning issuance you describe unless that kind of custom logic is part of the "script" that would be invoked as part of the notebook metadata extraction.  Does thamos provide the ability for that level of customization?

**Comment by pacospace** [comment]
> > create kernel with dependencies
> 
> What exactly does this mean? Using your "custom kernel" from above - `python3new` - in what manner are you altering the directory `/opt/app-root/src/.local/share/jupyter/kernels/python3new`? Could you provide its `kernel.json` and describe any other alterations, additional files, added into that directory's hierarchy?

I use jupyterlab-requirements [1] to create the custom kernel for my notebook:

- use the UI to add packages I want
- ask install those dependecies and I select a custom name for my kernel
in the background:
1. a new virtualenv is created for python.
2. dependencies are locked and a Pipfile.lock is created (either optimized using thoth reoslution engine [2] or pipenv. (user will be able to select this in the future)
3. micropipenv [3] install the dependencies in the virtualenv
4. run`ipython kernel install --user --name={kernel_name} --display-name 'Python ({kernel_name})'
5. assign the new kernel to the notebook of the user which is ready and using a kernel optimized for her/his own needs.
6. dependencies and locked dependencies + runtime environment + resolution engine are stored in notebook metadata.

you can now share that notebook and recreate that same kernel everywhere if you use that extension.

`/opt/app-root/src/.local/share/jupyter/kernels/python3new` is created when running step 4., a new kernel.json is created for the new kernel there which is a standard one:

![Screenshot from 2021-03-05 19-05-32](https://user-images.githubusercontent.com/27498679/110155378-d174af00-7de5-11eb-8d5d-b8fb198ff38c.png)

> 
> Regarding propagation of the notebook's requirements metadata into the container, there's a bit of a chicken/egg issue here. I wonder if we could introduce a "pre-step" during the node's processing in the bootstrapper sequence that extracts these custom requirements from the notebook's metadata and ensures their installation prior to the notebook's execution?

I think it could be done  in that way in kfp-notebook adding one two more steps in the operations performed.

> 
> I'm not sure if this step would be done against its own virtual env or not. I also don't see how we could entertain the kind of version checking/warning issuance you describe unless that kind of custom logic is part of the "script" that would be invoked as part of the notebook metadata extraction. Does thamos provide the ability for that level of customization?

Yes thamos library [3] uses a configuration file called `.thoth.yaml` [4] (also stored in notebook metadata, therefore it can be extracted together with the dependencies). 

for example I have a fedora 32 on my laptop and the `.thoth.yaml` file present here was created for rhel8, you can see the check that thamos can do:

```
thamos check

2021-03-05 19:09:29,253 [1398494] INFO     thamos.discover: No CUDA version detected
2021-03-05 19:09:29,294 [1398494] INFO     thamos.discover: Detected CPU: cpu_family: 6, cpu_model: 142, cpu_model_name: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ Message                                    ┃ Runtime environment ┃ Type      ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ Operating system name stated in the        │ rhel:8              │ ❌ ERROR  │
│ configuration file ('rhel') does not match │                     │           │
│ the one detected ('fedora')                │                     │           │
│ CPU model stated in the configuration file │ rhel:8              │ ⚠️ WARNING │
│ (None) does not match the one detected (6) │                     │           │
│ CPU model stated in the configuration file │ rhel:8              │ ⚠️ WARNING │
│ (None) does not match the one detected     │                     │           │
│ (142)                                      │                     │           │

```

[1] https://pypi.org/project/jupyterlab-requirements/
[2] https://thoth-station.ninja/
[3] https://pypi.org/project/micropipenv/
[4] https://github.com/thoth-station/thamos
[5] https://github.com/thoth-station/thamos#using-custom-configuration-file-template

**Comment by kevin-bates** [comment]
I see, so the ONLY difference with your "new kernel" and any other python kernel is which `python` command is used (and thus what env it's running within).  So, if the image has the appropriate environment there's really no need to propagate _this_ custom kernel since the only _customization_ is purely to ensure the correct python is invoked.

Had there been other customizations within the `kernel.json` file, there would be more to do with respect to ensuring the kernel customization is present in the image, but I think we're okay here.  Of course, there are still the env customizations that to talk about, but I'll defer to @akchinSTC as he's been more involved on that front.
